lib/auth/saml_test.go

```diff
@@ -87,3 +93,98 @@ func (s *SAMLSuite) TestCreateSAMLUser(c *check.C) {
+				fts9pMp4iS3tQguRc28gPdDo/T6VrJTVYUfUUsNDRtIrlB5O9igqqLnuaY9eqGi4
+				PUx0G0wRYJpRywoj8G0IkpfQTiX+CAC7dt5ws7ZrnGqCNBLGi5bGsaMmptVbsSEp
+				1TenntF54V1iR49IV5JqDhm1S0HmkleoJzKdc+6sP/xNepz9PJzuF9d9NubTLWgB
+				sK28YItcmWHdHXD/ODxVaehRjwIDAQABoyAwHjAOBgNVHQ8BAf8EBAMCB4AwDAYD
+				VR0TAQH/BAIwADANBgkqhkiG9w0BAQsFAAOCAQEAAVU6sNBdj76saHwOxGSdnEqQ
+				o2tMuR3msSM4F6wFK2UkKepsD7CYIf/PzNSNUqA5JIEUVeMqGyiHuAbU4C655nT1
+				IyJX1D/+r73sSp5jbIpQm2xoQGZnj6g/Kltw8OSOAw+DsMF/PLVqoWJp07u6ew/m
+				NxWsJKcZ5k+q4eMxci9mKRHHqsquWKXzQlURMNFI+mGaFwrKM4dmzaR0BEc+ilSx
+				QqUvQ74smsLK+zhNikmgjlGC5ob9g8XkhVAkJMAh2rb9onDNiRl68iAgczP88mXu
+				vN/o98dypzsPxXmw6tkDqIRPUAUbh465rlY5sKMmRgXi2rUfl/QV5nbozUo/HQ==
+			</X509Certificate>
+		</X509Data>
+	</KeyInfo>
+</KeyDescriptor>
+<AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
+		Location="https://my.site/saml/acs"/>
+</SPSSODescriptor>
+</EntityDescriptor>`
+
+func (s *SAMLSuite) TestEncryptedSAML(c *check.C) {
+	connector := services.NewSAMLConnector("spongebob", types.SAMLConnectorSpecV2{})
+
+	connector.SetEntityDescriptor(EntityDescriptor)
+	connector.SetIssuer("nil")
+	connector.SetSSO("nil")
+	connector.SetAssertionConsumerService("nil")
+
+	signingKeypair := &types.SigningKeyPair{
+		Cert:       fixtures.SigningCertPEM,
+		PrivateKey: fixtures.SigningKeyPEM,
+	}
+	connector.SetSigningKeyPair(signingKeypair)
+
+	encryptionKeypair := &types.EncryptionKeyPair{
+		Cert:       fixtures.EncryptionCertPEM,
+		PrivateKey: fixtures.EncryptionKeyPEM,
+	}
+	connector.SetEncryptionKeyPair(encryptionKeypair)
+
+	cert, err := tls.X509KeyPair([]byte(signingKeypair.Cert), []byte(signingKeypair.PrivateKey))
+	c.Assert(err, check.IsNil)
+
+	leaf, err := x509.ParseCertificate(cert.Certificate[0])
+	c.Assert(err, check.IsNil)
+
+	signingKeyStore, err := utils.ParseSigningKeyStorePEM(signingKeypair.PrivateKey, signingKeypair.Cert)
+	c.Assert(err, check.IsNil)
+
+	encryptionKeyStore, err := utils.ParseSigningKeyStorePEM(encryptionKeypair.PrivateKey, encryptionKeypair.Cert)
+	c.Assert(err, check.IsNil)
+
+	certStore := dsig.MemoryX509CertificateStore{
+		Roots: []*x509.Certificate{leaf},
+	}
+
+	provider := &saml2.SAMLServiceProvider{
+		IdentityProviderSSOURL:      "",
+		IdentityProviderIssuer:      "",
+		ServiceProviderIssuer:       "http://example.com/saml/acs/example",
+		AssertionConsumerServiceURL: "http://sp.example.com/demo1/index.php?acs",
+		SignAuthnRequests:           true,
+		AudienceURI:                 "http://example.com/saml/acs/example",
+		IDPCertificateStore:         &certStore,
+		SPKeyStore:                  encryptionKeyStore,
+		SPSigningKeyStore:           signingKeyStore,
+	}
+
+	response, err := provider.ValidateEncodedResponse(EncryptedResponse)
+	c.Assert(err, check.IsNil)
+
+	assertionInfo, err := RetrieveAssertionInfo(provider, connector, response)
+	c.Assert(err, check.IsNil)
+	if len(assertionInfo.Assertions) == 0 {
+		c.Error("no assertion found")
+	}
+}
```

Comment:
This could include a couple more tests for boundary conditions - e.g. zero assertions in response, missing subject, name, etc.

Also, do you know whether it's possible to have both unencrypted and encrypted assertions? since it would make sense to test the behavior.

Reply:
gosaml2 lacks good support for multiple assertions via the assertioninfo interface we use nor does it really make sense since you'll always have one on the other but never both for auth.

Reply:
Either there's one encrypted or one unencrypted and nothing else.

Reply:
Also gosaml2 validates the rest for us.

Reply:
Strictly speaking the SAML standard allows multiple assertions but this is pretty much never done in practice and most code including our current code and gosaml2 relies on there being one.

Reply:
Missed it but I added a call to have gosaml2 validate the decrypted assertions.